stdlib/camlinternalFormat.ml

```diff
@@ -1456,34 +1459,35 @@ let convert_int64 iconv n =
 (* Convert a float to string. *)
 (* Fix special case of "OCaml float format". *)
 let convert_float fconv prec x =
-  match snd fconv with
-  | Float_h | Float_H ->
+  let hex () =
     let sign =
       match fst fconv with
       | Float_flag_p -> '+'
       | Float_flag_s -> ' '
       | _ -> '-' in
-    let str = hexstring_of_float x prec sign in
-    begin match snd fconv with
-    | Float_H -> String.uppercase_ascii str
-    | _ -> str
-    end
-  | _ ->
+    hexstring_of_float x prec sign in
+  let caml_special_val s = match classify_float x with
+    | FP_normal | FP_subnormal | FP_zero -> s
+    | FP_infinite -> if x < 0.0 then "neg_infinity" else "infinity"
+    | FP_nan -> "nan" in
+  match snd fconv with
+  | Float_h -> hex ()
+  | Float_H -> String.uppercase_ascii (hex ())
+  | Float_CF -> caml_special_val (hex ())
+  | Float_F ->
     let str = format_float (format_of_fconv fconv prec) x in
-    if snd fconv <> Float_F then str else
+    let str =
```

Comment:
The whole `match snd fconv with ...` match would be more readable if the logic of this big `let str` block was moved into a function definition at the same place as `let hex ()`. (I wondered about suggesting to move both to outside the function, but they would need better name and the current local placement is probably fine.) What about `maybe_add_dot` or `add_dot_if_needed` as a function name (`string -> string`)?

Reply:
This does enlarge the diff but for the better. Done